nsx/resource_nsx_edge_dhcp.go

```diff
@@ -0,0 +1,358 @@
+			"cidr": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     false,
+				ValidateFunc: validateCidr,
+			},
+			"default_gw": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     false,
+				ValidateFunc: validateIP,
+			},
+			"ip_range": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     false,
+				ValidateFunc: validateIPRange,
+			},
+			"portgroup": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"edge_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+		},
+	}
+}
+
+func resourceNsxEdgeDHCPCreate(d *schema.ResourceData, meta interface{}) error {
+
+	dhcp, err := NewNsxEdgeDHCP(d)
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Creating NSX Edge DHCP: %#v", dhcp)
+
+	client := meta.(*govnsx.Client)
+
+	/* Get Edge configuration. Loop through all the vnics from 0-9.
+	   If the portgroup matches, add new address group for the existing vnic.
+	   else create a new vnic for the portgroup.
+	   If all the vnics are configured, return err */
+
+	edge := nsxresource.NewEdge(client)
+
+	edgeCfg, err := edge.Get(dhcp.edgeId)
+
+	if err != nil {
+		log.Printf("[Error] edge.Get() returned error : %v", err)
+		return err
+	}
+
+	// Update vnic configuration of Edge
+	vnics := []nsxtypes.Vnic{}
+	addrGroups := []nsxtypes.AddressGroup{}
+
+	addrGroup := nsxtypes.AddressGroup{}
+	addrGroup.PrimaryAddress = dhcp.vnicAddr
+	addrGroup.SubnetMask = dhcp.netMask
+	addrGroups = append(addrGroups, addrGroup)
+
+	vnic := nsxtypes.Vnic{}
+	vnic.Index = strconv.Itoa(0) //TODO:
+	vnic.PortgroupId = dhcp.portgroup
+	vnic.AddressGroups = addrGroups
+	vnic.IsConnected = true
+
+	vnics = append(vnics, vnic)
+
+	// Deploy appliance to true
+	edgeCfg.Appliances.DeployAppliances = true
+
+	edgeUpdateSpec := &nsxtypes.EdgeInstallSpec{
+		Tenant:     edgeCfg.Tenant,
+		Appliances: edgeCfg.Appliances,
+		Vnics:      vnics,
+	}
+
+	err = edge.Put(edgeUpdateSpec, dhcp.edgeId)
+
+	if err != nil {
+		log.Printf("[Error] edge.Put() returned error : %v", err)
+		return err
+	}
+
+	log.Printf("[INFO] Updated NSX Edge for DHCP configuration: %s", dhcp.edgeId)
+
+	// configure dhcp with the iprange and gw
+	ipPools := []nsxtypes.IPPool{}
+	ipPool := nsxtypes.IPPool{}
+	ipPool.IPRange = fmt.Sprintf(dhcp.ipRange.startIP + "-" + dhcp.ipRange.endIP)
+	ipPool.DefaultGw = dhcp.defaultGw
+	ipPool.SubnetMask = dhcp.netMask
+	ipPools = append(ipPools, ipPool)
+
+	dhcpConfigSpec := &nsxtypes.ConfigDHCPServiceSpec{
+		IPPools: ipPools,
+	}
+
+	edgeDHCP := nsxresource.NewEdgeDHCP(client)
+	err = edgeDHCP.Put(dhcpConfigSpec, dhcp.edgeId)
+
+	if err != nil {
+		log.Printf("[Error] edgeDHCP.Put() returned error : %v", err)
+		return err
+	}
+
+	d.SetId(dhcp.edgeId)
+	log.Printf("[INFO] Configured DHCP for NSX Edge: %s", dhcp.edgeId)
```

Comment:
We should set the location as Id and computed variable for DHCP edgeID though input variable edgeID and computed variable DhcpEdgeID remain same. However, user can change input edgeID. In that case we need do ForeNew.